File: carlolib/src/algorithm.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

int randint() {
    // randomly choose between 0 or 1 => map it to -1 1
    int rand_num = rand() % 2;
    int choice = 0;
    if (rand_num == 1) {
        choice = 1;
    } else {
        choice = -1;
    };
    return choice;
}

double* random_walk(int steps) {
    double SQRT_STEPS = sqrt((double)steps);
    // int random generator
    time_t t;
    srand((unsigned) time(&t));
    // allocate enough size for numbers
    double* numbers = malloc(sizeof(double) * steps);
    numbers[0] = 1;
    for (int i = 1; i < steps; i++) {
        int y = randint();
        numbers[i] = numbers[i-1] + (y / SQRT_STEPS); 
    }
    return numbers;
}

void free_arr(double* ptr) {
    free(ptr);
}

File: carlolib/algorithm.jl

function random_walk(steps::Int64)::Array{Float64, 1}
    SQRT_STEPS = sqrt(steps)
    walk = ones(steps)
    for idx in 2:steps
        if rand() < .5
            y = -1
        else
            y = 1
        end
        walk[idx] = walk[idx-1] + (y / SQRT_STEPS)
    end
    walk
end

random_walk(100)

multiple = 1
for i in 1:8
    global multiple
    t = @elapsed random_walk(multiple)
    println("jl_random_walk\t$t")
    multiple *= 10
end